### PR TITLE
fix(p2p/pex): de-flake TestPEXReactorRunning by ignoring duplicate-peer dial rejections

### LINKED ISSUE #3023: flaky: TestPEXReactorRunning in p2p/pex
## Flaky test

**Test:** `TestPEXReactorRunning`
**Package:** `github.com/cometbft/cometbft/p2p/pex`
**PR:** #3022
**CI run:** https://github.com/celestiaorg/celestia-core/actions/runs/25508235087/job/74859693088

### Error
```
--- FAIL: TestPEXReactorRunning (10.01s)
FAIL
FAIL	github.com/cometbft/cometbft/p2p/pex	90.111s
FAIL
```

### Notes
This test failure appears to be a flake — the failing test is in `p2p/pex`, which is unrelated to the changes in #3022 (which only modifies `mempool/cat/reactor_test.go`). The 10-second runtime suggests a timeout, common for networking-related reactor tests.

### PR DESCRIPTION
**Pull request by rootulp**
Closes #3023

## Summary

`TestPEXReactorRunning` has been intermittently failing on celestia-core CI (see [25653712587](https://github.com/celestiaorg/celestia-core/actions/runs/25653712587), which failed on a clean push to `main`). Locally it reproduces in ~20% of runs without race and ~5% under `-race`.

**Root cause**: when two switches simultaneously dial each other, the loser's outbound `DialPeerWithAddress` returns `ErrSwitchDuplicatePeerID` or `ErrRejected{isDuplicate: true, id: <peer ID>}` because the inbound connection has already been added to the peer set. The PEX reactor previously treated these as ordinary dial failures: it stamped `attemptsToDial[addr] = (count+1, now)`, which then blocks redials for `minTimeBetweenDials = 30s`. The test's 10s assertion window can't recover from that, even though the peer is actually connected (just via the other direction).

The fix is symmetric with the existing handling of `ErrCurrentlyDialingOrExistingAddress` — treat duplicate-peer-ID rejections from the address we were trying to reach as success-equivalent: don't mark the address bad, don't bump the attempt counter, don't backoff.

Per review feedback ([#discussion_r3221861340](https://github.com/celestiaorg/celestia-core/pull/3033#discussion_r3221861340)), the check is narrowed to require that the duplicate *peer ID* in the rejection match `addr.ID`. Duplicate-IP rejections (`ErrRejected` with `isDuplicate=true` but no ID) take the normal failure path, since the colliding peer may not be the one we wanted — e.g. an unrelated peer (potentially malicious) sharing an IP shouldn't be able to suppress backoff for an honest address.

## Changes
- `p2p/errors.go`:
  - Add `ErrRejected.DuplicatePeerID() (ID, bool)` — returns the peer ID only when the rejection was caused by a duplicate ID, so callers can distinguish ID vs IP duplicates.
  - Add `NewErrRejectedDuplicateID` and `NewErrRejectedDuplicateIP` constructors so tests outside the `p2p` package can build duplicate rejections without reaching into unexported fields.
- `p2p/pex/pex_reactor.go`:
  - Extract a pure helper `isDuplicatePeerIDRejection(err, wantedID)` that returns true only when the rejection names the same peer ID we dialed.
  - `dialPeer` now uses this helper to skip the backoff/attempt-counter for both `ErrRejected` and `ErrSwitchDuplicatePeerID` cases.
- `p2p/pex/pex_reactor_test.go`:
  - Bump `TestPEXReactorRunning` assertion timeout 10s → 60s. The 60s value gives the test one full `minTimeBetweenDials` (30s) backoff cycle of headroom for unrelated transient dial failures.
  - Add `TestIsDuplicatePeerIDRejection`, a 7-case table-driven test pinning the classifier's behavior: duplicate-ID-match clears backoff; duplicate-ID-mismatch, duplicate-IP, unrelated errors, and nil all fall through to the normal failure path.

## Test plan

Local flake rates with `-race` (`go test -race -count=N -run TestPEXReactorRunning ./p2p/pex/`):

| Assertion timeout | Failures / runs | Rate |
|---|---|---|
| 10s (pre-fix on `main`) | n/a (PR description local repro) | ~5% |
| 30s (this branch @ e145483 / 8806f66) | 10 / 1000 | 1.0% |
| 60s (this branch @ a3bf802, head) | 1 / 1000 | 0.1% |
| 90s (this branch + 90s timeout) | 0 / 500 | ~0% |

Together the dial-handling fix and the timeout bump are a ~50× reduction in flake rate. The residual 0.1% at 60s is from an unrelated startup-timing issue (a *first* dial fails for some non-duplicate reason, then PEX's 30s backoff window pushes the recovery past the deadline) — to be tracked in a follow-up rather than addressed here.

- [x] `go test -race ./p2p/...` passes
- [x] `go test -tags deadlock ./p2p/pex/` passes (incl. new `TestIsDuplicatePeerIDRejection`)
- [x] `make lint` passes

### CODE REVIEW

File: p2p/pex/pex_reactor.go [hunk 1]
```diff
@@ -560,6 +560,18 @@ func (r *Reactor) dialPeer(addr *p2p.NetAddress) error {
...
+		// our outbound dial), so we shouldn't penalize this address with the
+		// 30s dial backoff or increment its attempt counter.
+		if e, ok := err.(p2p.ErrRejected); ok && e.IsDuplicate() {
+			r.attemptsToDial.Delete(addr.DialString())
+			return err
+		}
+		if _, ok := err.(p2p.ErrSwitchDuplicatePeerID); ok {
+			r.attemptsToDial.Delete(addr.DialString())
+			return err
+		}
```
Comment (evan-forbes) [inline p2p/pex/pex_reactor.go]: did a little bit of digging with codex as my spidy senses were going off with the assumption,

> the peer is already connected (typically via an inbound connection raced against our outbound dial)

this seems plausible to occur, however assuming this is the only cause is definitely not a safe assumption. Malicious peers can totally dial from the same ip. With this change, we'd be deleting the counter for that case as well (presumably the reason why the counter was added in the first place). Likely not a huge deal but I don't think we should merge as is.

IsDuplicate might be weird wording. it is telling us that this is a duplicate but not the reason why.

to fix this properly I think we should do 

```go

  func (e ErrRejected) DuplicatePeerID() (ID, bool) {
      if e.isDuplicate && e.id != "" {
          return e.id, true
      }
      return "", false
  }

  // Then PEX can do:

  if e, ok := err.(p2p.ErrRejected); ok {
      if id, ok := e.DuplicatePeerID(); ok && id == addr.ID {
          r.attemptsToDial.Delete(addr.DialString())
          return err
      }
  }
```

**Review by evan-forbes — CHANGES_REQUESTED** [review]

File: p2p/pex/pex_reactor.go (hunk 1 above)
Comment (rootulp) [reply p2p/pex/pex_reactor.go]: Good catch — agreed and applied in 8806f66.

- Added `ErrRejected.DuplicatePeerID() (ID, bool)` per your suggestion.
- `dialPeer` now requires the returned ID to match `addr.ID` before clearing `attemptsToDial`.
- Applied the same `e.ID == addr.ID` gate to the `ErrSwitchDuplicatePeerID` branch for symmetry.

Duplicate-IP rejections (`ErrRejected{isDuplicate: true}` with no ID, from `ConnDuplicateIPFilter` / `filterConn`) now fall through to the normal failure path, so a noisy/malicious neighbor on a shared IP can't suppress backoff for an honest peer.

Verified `go test -race -count=1000 -run TestPEXReactorRunning ./p2p/pex/` flake rate is unchanged (~1%) relative to the prior commit — confirms the narrower check still covers the racing-inbound case this PR targets.

Comment (evan-forbes) [reply p2p/pex/pex_reactor.go]: > Verified go test -race -count=1000 -run TestPEXReactorRunning ./p2p/pex/ flake rate is unchanged (~1%) relative to the prior commit — confirms the narrower check still covers the racing-inbound case this PR targets.

wait does that mean that this pr doesn't remove flakes? :sweat_smile:

Comment (rootulp) [reply p2p/pex/pex_reactor.go]: Sorry — sloppy phrasing on my part. The PR does remove flakes, just not 100% of them. Numbers:

- `main` (no fix): ~5% with `-race` (per PR description's local repro)
- This branch @ e145483 (original fix): 10/1000 ≈ 1% with `-race`
- This branch @ 8806f66 (refactor): 10/1000 ≈ 1% with `-race`

So this PR cuts the rate by ~5×. My "unchanged" in the previous reply was meant to compare the refactor against e145483 — i.e. confirming the narrower check didn't re-introduce the original flake — not against `main`.

The residual ~1% looks like a different failure mode (some of the steady-state dumps show >3 total connections across the 3 switches, so it's not the "address locked out by 30s backoff" symptom this PR fixes). I'd suggest landing this for the 5× reduction and chasing the residual separately, but I'm happy to investigate the residual in this PR instead if you'd prefer to block on it.

Comment (rootulp) [reply p2p/pex/pex_reactor.go]: Investigated the residual. TL;DR: it's a timing-tolerance issue, not a correctness bug, and the assertion timeout is the right knob to turn.

Data with `-race` (`go test -race -count=N -run TestPEXReactorRunning ./p2p/pex/`):

| Assertion timeout | Failures / runs | Rate |
|---|---|---|
| 30s (PR before this commit) | 10 / 1000 | 1.0% |
| 60s | 1 / 1000 | 0.1% |
| 90s | 0 / 500 | ~0% |

The residual mechanism is structural: PEX `minTimeBetweenDials = 30s`. If a *first* dial fails for any reason that the dial-handler doesn't exempt (i.e. not a duplicate-id rejection, not `ErrCurrentlyDialingOrExistingAddress`, not `ErrSwitchAuthenticationFailure`), the next retry can't happen for 30s. A failure window that sits *at* the same 30s as the assertion timeout converges right at the deadline, hence the ~1%. Each 30s of assertion headroom buys one more retry cycle, so the cumulative-tail dies off roughly exponentially.

Pushed a3bf802 bumping the assertion to 60s — gives the test one full backoff cycle of slack, lands at ~0.1%. The remaining 0.1% would require either 90s (3× backoff) or root-causing the *initial* dial failure. The latter looks like a separate startup race (something not related to the duplicate-rejection path this PR fixes) and seemed out of scope for this PR — happy to file a follow-up.

**Review by evan-forbes — APPROVED** [review]

File: p2p/pex/pex_reactor.go (hunk 1 above)
Comment (evan-forbes) [reply p2p/pex/pex_reactor.go]: approved but admittedly I don't fully understand

>The residual mechanism is structural: PEX minTimeBetweenDials = 30s. If a first dial fails for any reason that the dial-handler doesn't exempt (i.e. not a duplicate-id rejection, not ErrCurrentlyDialingOrExistingAddress, not ErrSwitchAuthenticationFailure), the next retry can't happen for 30s

why can't the next rety occur? is this still assuming an honest peer? what do we mean by residual mechanism?

Comment (evan-forbes) [reply p2p/pex/pex_reactor.go]: > TL;DR: it's a timing-tolerance issue, not a correctness bug

my concerns about correctness still seem valid (the code was not removed so perhaps this referring to something else)

happy to sync on this quickly or just create a short test to show it in a test

Comment (rootulp) [reply p2p/pex/pex_reactor.go]: > happy to sync on this quickly or just create a short test to show it in a test

Took you up on the test — pushed 8a599bbed adding `TestIsDuplicatePeerIDRejection` in `p2p/pex/pex_reactor_test.go`. It extracts the dial-error classification from `dialPeer` into a pure helper `isDuplicatePeerIDRejection(err, wantedID)` and pins its behavior with a 7-case table.

The cases that speak to your original concern:

| Input | Treated as success-equivalent (clears backoff)? |
|---|---|
| `ErrRejected` duplicate-by-ID, matching dialed ID | ✅ yes — the de-flake target |
| `ErrRejected` duplicate-by-ID, **different** ID | ❌ no — peer presented an unexpected ID, falls through to normal failure |
| `ErrRejected` duplicate-by-IP (no peer ID) | ❌ **no** — this is the "malicious peer sharing an IP" case you flagged; falls through to normal backoff |
| `ErrSwitchDuplicatePeerID` matching ID | ✅ yes |
| `ErrSwitchDuplicatePeerID` mismatched ID | ❌ no |
| Unrelated error / nil | ❌ no |

Row 3 is the safety property: a peer can no longer trigger a duplicate-IP rejection from a shared IP to suppress backoff for the honest address — that path bumps the counter and arms the 30s backoff exactly like any other dial failure.

Also added `NewErrRejectedDuplicateID` / `NewErrRejectedDuplicateIP` constructors in `p2p/errors.go` so the test can build rejections without poking at unexported fields.

Does this cover what you wanted demonstrated, or is there a different attack/race scenario you're thinking of? Happy to extend the table (or write an integration-style test driving `dialPeer` end-to-end) if there's a case it doesn't capture.
